semgrep-core/src/fixing/Autofix_printer.ml

```diff
@@ -0,0 +1,128 @@
+ * WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the file
+ * LICENSE for more details.
+ *)
+
+open Common
+open AST_generic
+module MV = Metavariable
+
+(******************************************************************************)
+(* Handles AST printing for the purposes of autofix.
+ *
+ * The main printing logic happens elsewhere. This module's main purpose is to
+ * extend the existing AST printers so that they can avoid printing AST nodes
+ * which are lifted unchanged from either the target file (via metavariable
+ * bindings) or the fix pattern in the rule. This serves two purposes:
+ * - It lets us synthesize autofixes without having to implement printing for
+ *   all of the AST nodes in the fixed AST.
+ * - It lets us make minimal changes to target files by carrying over
+ *   formatting, comments, etc. from the original code.
+ *)
+(******************************************************************************)
+
+module PythonPrinter = Hybrid_print.Make (struct
+  class printer = Ugly_print_AST.python_printer
+end)
+
+let get_printer lang external_printer : Ugly_print_AST.printer_t option =
+  match lang with
+  | Lang.Python -> Some (new PythonPrinter.printer external_printer)
+  | _ -> None
+
+let original_source_of_ast source any =
+  let* start, end_ = Visitor_AST.range_of_any_opt any in
+  let starti = start.Parse_info.charpos in
+  let _, _, endi = Parse_info.get_token_end_info end_ in
+  let len = endi - starti in
+  let str = String.sub source starti len in
+  Some str
+
+(* Add each metavariable value to the lookup table so that it can be identified
+ * during printing *)
+let add_metavars tbl metavars =
+  List.iter
+    (fun (_, mval) ->
+      let any = MV.mvalue_to_any mval in
+      Hashtbl.replace tbl any `Metavar;
+      (* For each metavariable binding that is a list of things, we need to
+       * iterate through and add each item in the list to the table as well.
+       *
+       * For example, if `$...BAR` is bound to the separate arguments `1` and
+       * `2`, in the example below, then the complete argument list `(1, 2)`
+       * would never appear in the resulting AST that we attempt to print.
+       * Despite that, we would like to reuse the original text for `1` and `2`.
+       *
+       * foo($...BAR, 5) -> foo(1, 2, 5)
+       *
+       * We don't need to recurse any deeper, because individual list items are
+       * the smallest components of a metavariable value that will be lifted
+       * into the resulting AST.
+       * *)
+      match mval with
+      | MV.Args args ->
+          List.iter (fun arg -> Hashtbl.replace tbl (Ar arg) `Metavar) args
+      (* TODO iterate through other metavariable values that are lists *)
+      | _ -> ())
+    metavars
+
+(* Add each AST node to the lookup table so that it can be identified during
+ * printing *)
+let add_ast_nodes tbl ast =
```

Comment:
This name surprised me---I didn't expect `add_ast_nodes` to be the function that checked for patterns. This seems to actually be `add_fix_pattern_nodes`?

Reply:
Oh I am once again confused because I thought this was going through the whole target AST 🤦. I don't know why I'm so stuck on that.

Reply:
I will add some better comments here.